File: android/src/main/kotlin/com/oneplusdream/flutter_aws_chime/DataMessageObserver.kt
package com.oneplusdream.flutter_aws_chime

import com.amazonaws.services.chime.sdk.meetings.audiovideo.video.VideoTileState
import com.amazonaws.services.chime.sdk.meetings.realtime.datamessage.DataMessage
import com.amazonaws.services.chime.sdk.meetings.realtime.datamessage.DataMessageObserver

class DataMessageObserver(val methodChannel: MethodChannelCoordinator) : DataMessageObserver {
    override fun onDataMessageReceived(dataMessage: DataMessage) {
        methodChannel.callFlutterMethod(MethodCall.messageReceived, dataMessageToMap(dataMessage))
    }

    private fun dataMessageToMap(state: DataMessage): Map<String, Any?> {
        return mapOf(
                "attendeeId" to state.senderAttendeeId,
                "externalUserId" to state.senderExternalUserId,
                "topic" to state.topic,
                "timestampMs" to state.timestampMs,
                "message" to String(state.data),
                "throttled" to state.throttled
        )
    }
}

File: ios/Classes/DataMessageObserver.swift
//
//  DataMessageObserver.swift
//  flutter_aws_chime
//
//  Created by Conan on 2023/9/20.
//

import Foundation
import AmazonChimeSDK
import AmazonChimeSDKMedia
import AVFoundation
import Flutter
import UIKit

class MyDataMessageObserver: DataMessageObserver {
    func dataMessageDidReceived(dataMessage: AmazonChimeSDK.DataMessage) {
        methodChannel.callFlutterMethod(method: .messageReceived, args: messageInfoToDictionary(dataMessage:dataMessage))
    }
    
    var methodChannel: MethodChannelCoordinator

    init(withMethodChannel methodChannel: MethodChannelCoordinator) {
        self.methodChannel = methodChannel
    }

    private func messageInfoToDictionary(dataMessage: AmazonChimeSDK.DataMessage) ->[String:Any] {
        return [
            "attendeeId": dataMessage.senderAttendeeId,
            "externalUserId": dataMessage.senderExternalUserId,
            "topic":dataMessage.topic,
            "timestampMs":dataMessage.timestampMs,
            "message":  String(data: dataMessage.data, encoding: .utf8),
            "throttled": dataMessage.throttled,
        ]
    }
}
